userprog/address_space.cc

```diff
@@ -101,10 +121,21 @@ void AddressSpace::loadSegment(Executable &exe, uint32_t virtualAddr, uint32_t s
 /// Nothing for now!
 AddressSpace::~AddressSpace() {
     for (unsigned i = 0; i < numPages; i++) {
-        memoryMap->Clear(pageTable[i].physicalPage);
+        if (pageTable[i].valid) {
+            memoryMap->Clear(pageTable[i].physicalPage);
+        }
     }
 
     delete[] pageTable;
+    delete executable_file;
+
+#ifdef USE_SWAP
+    if (!fileSystem->Remove(swapFileName)) {
+        DEBUG('a', "Error removing swap file %s\n", swapFileName);
+    }
+
+    delete swapFile;
+#endif
```

Comment:
question: how can we guarantee swap files are removed properly when Nachos halts?

Currently, we're not properly cleaning up running threads when the machine halts. This means threads are not destroyed, which means spaces are not destroyed, which means the swap file is ultimately not removed.

Perhaps this can come as a followup improvement down the line when we focus on cleaning up memory leaks.